# Проект SQL "Анализ БД сервиса для чтения книг по подписке."

## Описание проекта
Стартаперы хотят запустить новое приложение для тех, кто любит читать. Необходимо проанализировать БД крупного сервиса для чтения книг по подписке. База данных содержит информацию о книгах, издательствах, авторах, а также пользовательские обзоры книг. На основе этих данных нужно сформулировать ценностное предложение для нового продукта.

**Основные задачи**: 
- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры; 
- Определите автора с самой высокой средней оценкой книг  - учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

## Описание данных 

**Таблица books** 

Содержит данные о книгах 
 
**Таблица authors**

Содержит данные об авторах
  
**Таблица publishers**

Содержит данные об издательствах

**Таблица ratings**

Содержит данные о пользовательских оценках книг
 
**Таблица reviews**

Содержит данные о пользовательских обзорах на книги:


## План работ

1. Установить соединение с БД.
2. Исследовать таблицы — вывести первые строки, посчитать количество строк в каждой таблице.
3. Написать по одному SQL-запросу для решения каждой задачи, вывести результаты.
     - Посчитайте, сколько книг вышло после 1 января 2000 года;
     - Для каждой книги посчитайте количество обзоров и среднюю оценку;
     - Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры; 
     - Определите автора с самой высокой средней оценкой книг  - учитывайте только книги с 50 и более оценками;
     - Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.
4. Сформулировать выводы по каждой из решённых задач.

## 1. Установка соединения с БД

In [1]:
# импортируем библиотеки
import pandas as pd
import sqlalchemy as sa

In [2]:
# устанавливаем параметры
db_config = {
'user': 'praktikum_student', # имя пользователя
'pwd': '****', # пароль
'host': '****',
'port': **, # порт подключения
'db': '*****' # название базы данных
}
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

In [3]:
# сохраняем коннектор
engine = sa.create_engine(connection_string, connect_args={'sslmode':'require'})

In [4]:
#функция для выполнения SQL-запроса
def show(sql):
    return pd.io.sql.read_sql(sql, con = engine) 

In [5]:
# добавим настройку, чтобы текст в столбцах отражался полностью
pd.options.display.max_colwidth = 1000

## 2. Исследование таблиц — вывести первые строки, посчитать количество строк в каждой таблице.

### Таблица books

In [6]:
query_books = '''SELECT * FROM books LIMIT 5'''
     
books = show(query_books)

In [7]:
books

book_id  author_id                                                  title  \
0        1        546                                           'Salem's Lot   
1        2        465                     1 000 Places to See Before You Die   
2        3        407    13 Little Blue Envelopes (Little Blue Envelope  #1)   
3        4         82  1491: New Revelations of the Americas Before Columbus   
4        5        125                                                   1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [8]:
query_books = '''SELECT count(*) FROM books'''
     
books = show(query_books)

In [9]:
books

count
0   1000

### Таблица authors

In [10]:
query_authors = '''SELECT * FROM authors LIMIT 5'''
     
authors = show(query_authors)

In [11]:
authors

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [12]:
query_authors = '''SELECT count(*) FROM authors LIMIT 5'''
     
authors = show(query_authors)

In [13]:
authors

count
0    636

### Таблица publishers

In [14]:
query_publishers = '''SELECT * FROM publishers LIMIT 5'''
     
publishers = show(query_publishers)

In [15]:
publishers

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [16]:
query_publishers = '''SELECT count(*) FROM publishers LIMIT 5'''
     
publishers = show(query_publishers)

In [17]:
publishers

count
0    340

### Таблица ratings

In [18]:
query_ratings = '''SELECT * FROM ratings LIMIT 5'''
     
ratings = show(query_ratings)

In [19]:
ratings

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [20]:
query_ratings = '''SELECT count(*) FROM ratings LIMIT 5'''
     
ratings = show(query_ratings)

In [21]:
ratings

count
0   6456

### Таблица  reviews

In [22]:
query_reviews = '''SELECT * FROM reviews LIMIT 5'''
     
reviews = show(query_reviews)

In [23]:
reviews

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                                                                                               text  
0                                        Mention society tell send professor analysis. Over provide race technology continue these.  
1                                                 Foot glass pretty audience hit themselves. Among admit investment argue security.  
2  Listen treat keep worry. Miss husband tax but person sport treatment industry. Kitchen decision deep the. Social party body the.  
3          Finally month interesting blue could nature cultural bit. Prepare beat finish grow that smile teach. Dream me play near.  
4                  Nation purpose heavy give wait song will. List dinner another whole positive radio fast. Music staff many green.

In [24]:
query_reviews = '''SELECT count(*) FROM reviews LIMIT 5'''
     
reviews = show(query_reviews)

In [25]:
reviews

count
0   2793

## 3. Решение задач

### 3.1 Посчитайте, сколько книг вышло после 1 января 2000 года;

In [26]:
count_book = '''
                 SELECT COUNT(book_id) AS "Кол-во книг"
                 FROM books 
                 WHERE publication_date > '2000-01-01'
             '''
     
count_book = show(count_book)

In [27]:
count_book

Кол-во книг
0          819

После 1 января 2000 года вышло 819 книг.

### 3.2 Для каждой книги посчитайте количество обзоров и среднюю оценку;

In [28]:
count_rat_rev = '''
                 SELECT b.title AS "Книга",
                        COUNT(DISTINCT rev.review_id) AS cnt_review,
                        ROUND(AVG(r.rating),2) AS avg_rating
                 FROM books AS b
                 LEFT JOIN  ratings AS r ON b.book_id=r.book_id
                 LEFT JOIN reviews AS rev ON b.book_id=rev.book_id
                 GROUP BY b.book_id, b.title
                 ORDER BY cnt_review DESC, avg_rating DESC
                '''
     
count_rat_rev = show(count_rat_rev)

In [29]:
#pd.set_option('display.max_rows', None)

In [37]:
count_rat_rev

Книга  cnt_review  \
0                                        Twilight (Twilight  #1)           7   
1    Harry Potter and the Prisoner of Azkaban (Harry Potter  #3)           6   
2     Harry Potter and the Chamber of Secrets (Harry Potter  #2)           6   
3                                                 The Book Thief           6   
4                                               The Glass Castle           6   
..                                                           ...         ...   
995         Disney's Beauty and the Beast (A Little Golden Book)           0   
996                                         Leonardo's Notebooks           0   
997                                    Essential Tales and Poems           0   
998              Anne Rice's The Vampire Lestat: A Graphic Novel           0   
999                                      The Natural Way to Draw           0   

     avg_rating  
0          3.66  
1          4.41  
2          4.29  
3          4.26  
4          4.21  
..          ...  
995        4.00  
996        4.00  
997        4.00  
998        3.67  
999        3.00  

[1000 rows x 3 columns]

Больше всего обзоров у книги "Twilight" - 7,  а ее рейтинг 3,66. Также Среди лидеров "Harry Potter and the Prisoner of Azkaban" и "Harry Potter and the Chamber of Secrets" с кол-вом обзоров - 6 и со средними оценками   4.41 и 4.29 соответственно. "The Book Thief"- 6 и рейтинг 4.26, "The Glass Castle" - 6 и рейтинг 4.21.

### 3.3 Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;

In [31]:
book_50 = '''
            SELECT p.publisher AS "Издательство",
                   COUNT(b.book_id) AS "Количество книг"
             FROM publishers AS p
             LEFT JOIN books AS b ON p.publisher_id = b.publisher_id
             WHERE b.num_pages > 50
             GROUP BY p.publisher_id
             ORDER BY COUNT(b.book_id) DESC
             LIMIT 1
       
           '''     
     
book_50 = show(book_50)

In [32]:
book_50

Издательство  Количество книг
0  Penguin Books               42

Больше всего книг выпустило издательство Penguin Books 42 книги.

### 3.4 Определите автора с самой высокой средней оценкой книг - учитывайте только книги с 50 и более оценками;

In [38]:
book_high_rating = '''
    WITH
    t1 AS   (SELECT b.book_id,
                   b.author_id AS author_id,
                   b.title,
                   AVG(r.rating) AS avg_rating
             FROM books AS b
             LEFT JOIN ratings AS r ON b.book_id=r.book_id
             GROUP BY b.book_id
             HAVING COUNT(r.rating)>=50)
             
   SELECT a.author AS "Автор",
          AVG(t1.avg_rating) AS "Средний рейтинг книг"
   FROM authors AS a
   JOIN t1 ON a.author_id = t1.author_id
   GROUP BY  a.author_id
   ORDER BY AVG(t1.avg_rating) DESC
   
       
           '''     
     
book_high_rating = show(book_high_rating)

In [39]:
book_high_rating

Автор  Средний рейтинг книг
0                           J.K. Rowling/Mary GrandPré              4.283844
1                    Markus Zusak/Cao Xuân Việt Khương              4.264151
2                                       J.R.R. Tolkien              4.258446
3                                    Louisa May Alcott              4.192308
4                                         Rick Riordan              4.080645
5                                      William Golding              3.901408
6                                        J.D. Salinger              3.825581
7             Paulo Coelho/Alan R. Clarke/Özdemir İnce              3.789474
8   William Shakespeare/Paul Werstine/Barbara A. Mowat              3.787879
9                                            Dan Brown              3.754540
10                                          Lois Lowry              3.750000
11             George Orwell/Boris Grabnar/Peter Škerl              3.729730
12                                     Stephenie Meyer              3.662500
13                                      John Steinbeck              3.622951

Автор с самой высокой средней оценкой книг J.K. Rowling.

### 3.5 Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [35]:
avg_review = '''
WITH 
  t1 AS (SELECT username,
               COUNT(rating_id)
        FROM ratings
        GROUP BY username
        HAVING  COUNT(rating) > 48
        ORDER BY COUNT(rating) DESC),
  
  t2 AS (SELECT username,
               COUNT(review_id) AS cnt_rev
        FROM reviews
        GROUP BY username
        ORDER BY AVG(review_id) DESC )
   
   SELECT AVG(cnt_rev) AS "Cреднее количество обзоров"
   FROM t1
   JOIN t2 ON t1.username = t2.username
   
       
           '''     
     
avg_review = show(avg_review)

In [36]:
avg_review

Cреднее количество обзоров
0                        24.0

Cреднее количество обзоров от пользователей, которые поставили больше 48 оценок 24.

## Вывод
Была проанализирована БД крупного сервиса для чтения книг по подписке.
Выявили, что 
 - После 1 января 2000 года вышло 819 книг.
 - Больше всего обзоров у книги "Twilight" - 7, а ее рейтинг 3,66. Также Среди лидеров "Harry Potter and the Prisoner of Azkaban" и "Harry Potter and the Chamber of Secrets" с кол-вом обзоров - 6 и со средними оценками 4.41 и 4.29 соответственно. "The Book Thief"- 6 и рейтинг 4.26, "The Glass Castle" - 6 и рейтинг 4.21.
 - Больше всего книг выпустило издательство Penguin Books 42 книги.
 - Автор с самой высокой средней оценкой книг J.K. Rowling.
 - Cреднее количество обзоров от пользователей, которые поставили больше 48 оценок 24.


На основе полученных данных сформулируем ценностное предложение для нового продукта:

1. Выводить в топ свежие книги (вышедшие с 2000 года и позднее);

2. Предлагать подписчикам книги, пользующиеся сейчас наибольшей популярностью.

3. Предлагать книги издательств Penguin Books, как лидера среди издательств. 

4. Привлекать внимание пользователей к книгам популярных авторов J.K. Rowling/Mary GrandPré, Markus Zusak,J.R.R. Tolkien,Louisa May Alcott и т.д.
